### PR TITLE
Optimize MF model

### PR DESCRIPTION
**Pull request by rayrayraykk**
~~Optimize MF model forward:~~

* ~~Before: (user_embedding * item_embedding) [mask]~~
* ~~Now: (user_embedding[user_mask].to_sparse() * item_embedding[item_mask].to_sparse()[mask]~~

Optimization of MF

1. Convert MF model to `nn.Embedding` Version: forward with used embeddings
2. Saving client data to `processed_data.pkl` to save time when repeating the exp.

### CODE REVIEW

File: federatedscope/mf/model/model.py [hunk 1]
```diff
@@ -15,47 +14,34 @@ class BasicMFNet(Module):
...
-        self.register_parameter('embed_user', self.embed_user)
-        self.embed_item = Parameter(
-            torch.normal(mean=0,
-                         std=0.1,
-                         size=(num_item, num_hidden),
-                         requires_grad=True,
-                         dtype=torch.float32))
-        self.register_parameter('embed_item', self.embed_item)
+        self.num_user, self.num_item = num_user, num_item
+        self.embed_user = torch.nn.Embedding(num_user, num_hidden, sparse=True)
```
Comment (joneswong) [inline federatedscope/mf/model/model.py:18]: I am wondering why sparse is needed?

File: federatedscope/mf/model/model.py [hunk 2]
```diff
@@ -15,47 +14,34 @@ class BasicMFNet(Module):
...
-                                        device=pred.device,
-                                        dtype=torch.float32).to_dense()
-        mask = torch.sparse_coo_tensor(indices,
-                                       np.ones(len(ratings)),
-                                       size=pred.shape,
-                                       device=pred.device,
-                                       dtype=torch.float32).to_dense()
-
-        return mask * pred, label, float(np.prod(pred.size())) / len(ratings)
+        device = self.embed_user.weight.device
```
Comment (joneswong) [inline federatedscope/mf/model/model.py:22]: what indices and ratings are should be stated

File: federatedscope/mf/model/model.py [hunk 3]
```diff
@@ -15,47 +14,34 @@ class BasicMFNet(Module):
...
+        indices = torch.tensor(np.array(indices)).to(device)
+
+        user_embedding = self.embed_user(indices[0])
+        item_embedding = self.embed_item(indices[1])
+
+        pred = torch.diag(torch.matmul(user_embedding, item_embedding.T))
+
+        label = torch.tensor(np.array(ratings)).to(device)
+
+        return pred, label
```
Comment (joneswong) [inline federatedscope/mf/model/model.py:33]: when do you calculate these for negative examples?

**Comment by joneswong** [comment]
@DavdGao hi dawei, please help us check the changes about datasets. Thanks!

File: federatedscope/mf/dataset/movielens.py [hunk 4]
```diff
@@ -75,7 +97,16 @@ def __init__(self, root, num_client, train_portion=0.9, download=True):
                                "You can use download=True to download it")
 
         ratings = self._load_meta()
-        self._split_n_clients_rating(ratings, num_client, 1 - train_portion)
+
+        self.processed_data = os.path.join(self.root, self.base_folder,
+                                           'processed_data.pkl')
+        if os.path.exists(self.processed_data):
```
Comment (DavdGao) [inline federatedscope/mf/dataset/movielens.py:103]: I don't think line 104-105 will be executed, since a new exp directory (`self.root`) will be created for each time.

File: federatedscope/mf/model/model.py (hunk 1 above)
Comment (DavdGao) [inline federatedscope/mf/model/model.py:18]: What's the difference between `torch.nn.Embedding` and `torch.normal`, and is `torch.nn.Embedding` much faster than `torch.normal`?

Comment (DavdGao) [reply federatedscope/mf/model/model.py:18]: Me too. In my opinion, the label matrix for MF task is very sparse, but the user/item embedding  and prediction results should be dense.

File: federatedscope/mf/model/model.py [hunk 5]
```diff
@@ -15,47 +14,34 @@ class BasicMFNet(Module):
...
-
-        return mask * pred, label, float(np.prod(pred.size())) / len(ratings)
+        device = self.embed_user.weight.device
+
+        indices = torch.tensor(np.array(indices)).to(device)
+
+        user_embedding = self.embed_user(indices[0])
+        item_embedding = self.embed_item(indices[1])
+
+        pred = torch.diag(torch.matmul(user_embedding, item_embedding.T))
```
Comment (DavdGao) [inline federatedscope/mf/model/model.py:29]: Why `pred` is calculated by `torch.diag` here?

**Review by DavdGao — COMMENTED** [review]
Please see the inline comments

File: federatedscope/mf/model/model.py (hunk 3 above)
Comment (joneswong) [reply federatedscope/mf/model/model.py:33]: It is OK to me to just fit the observed entries, but it is a very naive baseline approach I guess

**Review by joneswong — APPROVED** [review]
approved.
